pkg/services/sqlstore/migrations/ualert/ualert.go

```diff
@@ -876,3 +879,104 @@ func (c updateRulesOrderInGroup) Exec(sess *xorm.Session, migrator *migrator.Mig
 	}
 	return nil
 }
+
+func AddExternalAlertmanagerToDatasourceMigration(mg *migrator.Migrator) {
+	mg.AddMigration("migrate external alertmanagers to datsourcse", &externalAlertmanagerToDatasources{})
+}
+
+type externalAlertmanagerToDatasources struct {
+	migrator.MigrationBase
+}
+
+// Just copy old struct code to use Alertmanagers as new one is not getting it.
+type AdminConfiguration struct {
+	ID    int64 `xorm:"pk autoincr 'id'"`
+	OrgID int64 `xorm:"org_id"`
+
+	Alertmanagers []string
+
+	// SendAlertsTo indicates which set of alertmanagers will handle the alert.
+	SendAlertsTo ngmodels.AlertmanagersChoice `xorm:"send_alerts_to"`
+
+	CreatedAt int64 `xorm:"created"`
+	UpdatedAt int64 `xorm:"updated"`
+}
+
+func (e externalAlertmanagerToDatasources) SQL(dialect migrator.Dialect) string {
+	return "migrate external alertmanagers to datasource"
+}
+
+func (e externalAlertmanagerToDatasources) Exec(sess *xorm.Session, mg *migrator.Migrator) error {
+	var results []AdminConfiguration
+	err := sess.SQL("SELECT org_id, alertmanagers FROM ngalert_configuration").Find(&results)
+	if err != nil {
+		return err
+	}
+
+	for _, result := range results {
+		for _, am := range result.Alertmanagers {
+			u, err := url.Parse(am)
+			if err != nil {
+				return err
+			}
+			uri := fmt.Sprintf("%s://%s", u.Scheme, u.Host)
+
+			uid, err := generateNewDatasourceUid(sess, result.OrgID)
+			if err != nil {
+				return err
+			}
+			ds := &datasources.DataSource{
+				OrgId:   result.OrgID,
+				Name:    fmt.Sprintf("alertmanager-%s", uid),
+				Type:    "alertmanager",
+				Access:  "proxy",
+				Url:     uri,
+				Created: time.Unix(result.CreatedAt, 0),
+				Updated: time.Unix(result.UpdatedAt, 0),
+				Uid:     uid,
+				Version: 1,
+				JsonData: simplejson.NewFromAny(map[string]interface{}{
+					"handleGrafanaManagedAlerts": true,
+					"implementation":             "prometheus",
+				}),
+				SecureJsonData: map[string][]byte{},
+			}
+
+			if u.User != nil {
+				ds.BasicAuth = true
+				ds.BasicAuthUser = u.User.Username()
+				if password, ok := u.User.Password(); ok {
+					ds.SecureJsonData = GetEncryptedJsonData(map[string]string{
+						"basicAuthPassword": password,
+					})
+				}
+			}
+
+			rowsAffected, err := sess.Table("data_source").Insert(ds)
+			if err != nil {
+				return err
+			}
+			if rowsAffected == 0 {
+				return fmt.Errorf("expected 1 row, got %d", rowsAffected)
+			}
+		}
+	}
+
+	return nil
+}
+func generateNewDatasourceUid(sess *xorm.Session, orgId int64) (string, error) {
+	for i := 0; i < 3; i++ {
+		uid := util.GenerateShortUID()
+
+		exists, err := sess.Table("data_source").Where("uid = ? AND org_id = ?", uid, orgId).Exist()
```

Comment:
If `exists` is false after three attempts won't this return `"", nil`?

Reply:
If it is `false`, it will return a valid uid on the first attempt, right? Am I missing something?

Reply:
Apologies, I mean if `exists` is `true` after three attempts

Reply:
If `exists` is `true` after 3 attempts it will `return "", datasources.ErrDataSourceFailedGenerateUniqueUid`.

I am failing to see your point. The only thing we can do here, AFAIK, is to raise the number of attempts. right?

Reply:
Ah apologies, I somehow misread the return at the end of the function.